modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/Services/ServicesClient.cs

```diff
@@ -202,7 +215,123 @@ public IServicesClient WithServerKeepBinary()
+            private readonly string _srvcName;
+            
+            /** Ignite services. */
+            private readonly ServicesClient _srvc;
+
+            /** Flag of topology update progress. */
+            private int _updateInProgress;
+
+            /** Time of the last received topology. */
+            private long _lastUpdateRequestTime;
+            
+            /** UUIDs of the nodes with at least one service instance. */
+            private volatile IList<Guid> _nodes;
+
+            /** Last cluster topology version when current service topology was actual. */
+            private long _lastAffTop;
+
+            /** */
+            internal ServiceTopology(string name, ServicesClient srvc)
+            {
+                _srvcName = name;
+                _srvc = srvc;
+            }
+
+            /// <summary>
+            /// Asynchronously updates the service topology.
+            /// </summary>
+            private void UpdateTopologyAsync()
+            {
+                if (Interlocked.Exchange(ref _updateInProgress, 1) == 1)
+                    return;
+
+                var socket = _srvc._ignite.Socket;
+
+                var topVer = socket.GetTopologyVersion();
+
+                var log = _srvc._ignite.GetConfiguration().Logger;
+
+                var groupNodes = _srvc._clusterGroup?.GetNodes();
+                
+                _nodes = socket.DoOutInOpAsync(ClientOp.ServiceGetTopology,
+                    ctx => ctx.Writer.WriteString(_srvcName),
+                    ctx =>
+                    {
+                        var cnt = ctx.Reader.ReadInt();
+
+                        IList<Guid> res = new List<Guid>(cnt);
+
+                        for (var i = 0; i < cnt; ++i)
+                            res.Add(BinaryUtils.ReadGuid(ctx.Reader.Stream));
+
+                        Interlocked.Exchange(ref _lastUpdateRequestTime, DateTime.Now.Ticks);
+                        Interlocked.Exchange(ref _lastAffTop, topVer);
+                        Interlocked.Exchange(ref _updateInProgress, 0);
+
+                        var filteredTopology = FilterTopology(res, groupNodes?.Select(n => n.Id).ToList());
+
+                        log.Debug("Topology of service '" + _srvcName + "' has been updated. The " +
+                                  "service instance nodes: " + string.Join(", ", res.Select(gid=>gid.ToString())) +
+                                  ". Effective topology with the cluster group is: " + 
+                                  string.Join(", ", filteredTopology.Select(gid=>gid.ToString())) + '.');
+
+                        return filteredTopology;
+                    }, (status, err) =>
+                    {
+                        Interlocked.Exchange(ref _updateInProgress, 0);
+
+                        log.Error("Failed to update topology of the service '" + _srvcName + "'.", err);
+
+                        return _nodes;
+                    }).Result;
```

Comment:
Don't use `.Result` - it blocks the thread to wait for the Task. Prefer `await`.

Reply:
Yes, forgot the async